### PR TITLE
Convert launcher job and statefulset worker to pod

### PR DESCRIPTION
**Pull request by carmark**
fix https://github.com/kubeflow/mpi-operator/issues/201

need to make sure whether we need:
* BackoffLimit
* ActiveDeadlineSeconds

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/mpi-operator/203)

### CODE REVIEW

**Comment by carmark** [comment]
@terrytangyuan Could you please take a look? I move the previous patch to v1.

File: deploy/v1/mpi-operator.yaml [hunk 1]
```diff
@@ -0,0 +1,196 @@
+# --------------------------------------------------
+# - Single configuration deployment YAML for MPI-Operator
+# - Include:
+#      CRD deployment
+#      Namespace deployment
+#      RBAC deployment
+#      MPI-operator container deployment set
+# --------------------------------------------------
```
Comment (terrytangyuan) [inline deploy/v1/mpi-operator.yaml:8]: Why is this file in the diff? Perhaps do a rebase to fix this?

File: pkg/controllers/v1/mpi_job_controller_status.go [hunk 2]
```diff
@@ -91,14 +98,25 @@ func isFailed(status common.JobStatus) bool {
...
+
 // setCondition updates the mpiJob to include the provided condition.
 // If the condition that we are about to add already exists
 // and has the same status and reason then we are not going to update.
 func setCondition(status *common.JobStatus, condition common.JobCondition) {
 	// Do nothing if MPIJobStatus have failed condition
-	if isFailed(*status) {
-		return
-	}
+	//if isFailed(*status) {
```
Comment (terrytangyuan) [inline pkg/controllers/v1/mpi_job_controller_status.go:117]: Do we still need this?

File: pkg/apis/kubeflow/v1/constants.go [hunk 3]
```diff
@@ -14,7 +14,11 @@
 
 package v1
 
+import common "github.com/kubeflow/common/pkg/apis/common/v1"
+
 const (
 	// EnvKubeflowNamespace is ENV for kubeflow namespace specified by user.
 	EnvKubeflowNamespace = "KUBEFLOW_NAMESPACE"
+	// DefaultRestartPolicy is default RestartPolicy for TFReplicaSpec.
```
Comment (terrytangyuan) [inline pkg/apis/kubeflow/v1/constants.go:22]: Should not be TFReplicaSpec

File: cmd/mpi-operator.v1/app/server.go [hunk 4]
```diff
@@ -155,8 +155,7 @@ func Run(opt *options.ServerOption) error {
 			kubeInformerFactory.Core().V1().ServiceAccounts(),
 			kubeInformerFactory.Rbac().V1().Roles(),
 			kubeInformerFactory.Rbac().V1().RoleBindings(),
-			kubeInformerFactory.Apps().V1().StatefulSets(),
-			kubeInformerFactory.Batch().V1().Jobs(),
```
Comment (terrytangyuan) [inline cmd/mpi-operator.v1/app/server.go:159]: Update the deployment yaml files too?

File: Dockerfile [hunk 5]
```diff
@@ -2,9 +2,9 @@ FROM golang:1.13.6 AS build
...
 ADD . /go/src/github.com/kubeflow/mpi-operator
 WORKDIR /go/src/github.com/kubeflow/mpi-operator
-RUN go build -o mpi-operator github.com/kubeflow/mpi-operator/cmd/mpi-operator.v1alpha2
+RUN make
 
 FROM gcr.io/distroless/base-debian10:latest
-COPY --from=build /go/src/github.com/kubeflow/mpi-operator/mpi-operator /opt/
-ENTRYPOINT ["/opt/mpi-operator"]
+COPY --from=build /go/src/github.com/kubeflow/mpi-operator/_output/cmd/bin/mpi-operator* /opt/
+ENTRYPOINT ["/opt/mpi-operator.v1alpha2"]
```
Comment (terrytangyuan) [inline Dockerfile:9]: Should we point to v1 now?

File: deploy/v1/mpi-operator.yaml (hunk 1 above)
Comment (carmark) [reply deploy/v1/mpi-operator.yaml:8]: separate a new directory for different api version.

File: Dockerfile (hunk 5 above)
Comment (carmark) [reply Dockerfile:9]: By default, we use v1alpha2 as the default one, and override it in kubernetes deployment yaml file.

File: cmd/mpi-operator.v1/app/server.go (hunk 4 above)
Comment (carmark) [reply cmd/mpi-operator.v1/app/server.go:159]: already updated it.

**Comment by terrytangyuan** [comment]
@carmark There's already a yaml file in https://github.com/kubeflow/mpi-operator/blob/2b68b44e90fb90f6433329b5a57deb59d67bfcff/deploy/mpi-operator-v1.yaml

**Comment by terrytangyuan** [comment]
Also CI build failed: https://travis-ci.org/github/kubeflow/mpi-operator/builds/678854834

**Comment by carmark** [comment]
Thanks @terrytangyuan, I will update the unit test.

**Comment by terrytangyuan** [comment]
Could you also remove the duplicated manifest?

**Comment by carmark** [comment]
sure, will do it.

**Comment by terrytangyuan** [comment]
There are some conflicts after I merged another PR. Could you resolve them and once Travis build passes I think it's good to merge.

**Comment by terrytangyuan** [comment]
@carmark Build is still failing. Could you take another look? https://travis-ci.org/github/kubeflow/mpi-operator/pull_requests

**Comment by carmark** [comment]
@terrytangyuan thanks, could you please review it again? I fixed the conflict and unit test issues.

File: pkg/controllers/v1/mpi_job_controller.go [hunk 6]
```diff
@@ -801,44 +839,70 @@ func (c *MPIJobController) updateMPIJobStatus(mpiJob *kubeflow.MPIJob, launcher
...
+		evict   = 0
+	)
+
+	initializeMPIJobStatuses(mpiJob, kubeflow.MPIReplicaTypeWorker)
+	//spec := mpiJob.Spec.MPIReplicaSpecs[kubeflow.MPIReplicaTypeWorker]
+	for i := 0; i < len(worker); i++ {
+		switch worker[i].Status.Phase {
+		case corev1.PodFailed:
+			mpiJob.Status.ReplicaStatuses[common.ReplicaType(kubeflow.MPIReplicaTypeWorker)].Failed += 1
+			if worker[i].Status.Reason == "Evicted" {
```
Comment (alculquicondor) [inline pkg/controllers/v1/mpi_job_controller.go:878]: what if it failed for a different reason?

**Pull request merged**